### PR TITLE
Migrate BatchExporterProcessor to async

### LINKED ISSUE #5778: Async BatchExportProcessor
### Package

OpenTelemetry

### Is your feature request related to a problem?

BatchExportProcessor is unnecessarily spawning a thread for the export. This thread spends 99% of its time blocked. This is a little waste of resource and it can also mislead investigations about thread pool starvation (a blocked thread looks like an issue until you find out it's not a thread pool thread).

### What is the expected behavior?

The export spends its time waiting for a trigger or for the response of the export operation. It is completely fine using the .NET thread pool as long as all operations are asynchronous. This would make OTEL a better citizen in the .NET ecosystem.

The work would be:
- remove the [exporter thread](https://github.com/open-telemetry/opentelemetry-dotnet/blob/b15f9a82e3fe2052de08ab580428b4ff56f9dec6/src/OpenTelemetry/BatchExportProcessor.cs#L60-L64)
- migrate [BaseExporter.Export](https://github.com/open-telemetry/opentelemetry-dotnet/blob/b15f9a82e3fe2052de08ab580428b4ff56f9dec6/src/OpenTelemetry/BaseExporter.cs#L43) to async. Or probably add a new method to avoid a breaking change

### Which alternative solutions or features have you considered?

N/A

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by verdie-g**
Fixes #5778

## Changes

This makes BatchExporterProcessor async to avoid having dedicated threads that are idle 99% of the time.

I first tried to make the implementation close to the original synchronous implementation where only the ExportProc method was allowed to call `BaseExporter.ExportAsync` but it was a unreviewable deadlock fest. This proposed implementation allows any method (Export, Flush, Shutdown) to also call `ExportAsync` but always limit a single concurrent export. This greatly simplifies the code and only took me 30 minutes to implement.

It is very important that we first discuss about how `BaseExporter.ExportAsync` will be implemented on all available exporters before merging this change, otherwise, this is going to push synchronous job on the thread pool which will make the situation worse than before.

## Merge requirement checklist

* [ ] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [ ] Unit tests added/updated
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

**Comment by cijothomas** [comment]
Can you describe the overall idea on how this will be implemented? This a core functionality, and there cannot be any breaking changes at all. It is intentional that the current design spins up own thread and not rely on async tasks/thread-pool. 
And also the benefits of this change.

**Comment by verdie-g** [comment]
The exporter loop is kept but is now running on the thread pool using a simple `Task.Run`. The Export, Flush, and Shutdown operations were allowed to speed up the exporting by disabling the delay between two exports. In this change, it's done differently to simplify the code. The 3 operations are allowed to call the `BaseExporter.ExportAsync` method but only one concurrent export is allowed. `BaseExporter.ExportAsync` is a new method that by default does async-over-sync by calling `Export` just to avoid a breaking. Though, it is important that this method gets overridden by the different implementations (zipkin, OTEL) to avoid pushing synchronous operations on the thread pool.

The idea of a thread pool is to avoid the cost of the creation/deletion of a thread as well the cost of the resources associated to the thread. But most importantly, it avoids having to perform a context switch to yield the execution to another task. In an ideal application, the thread pool has as many thread as CPU cores, and each thread never leaves its core. Each context switch will impact the throughput of the thread pool. If there is only a few extra threads used, it's fine, but if all libraries start spawning dedicated threads for their background jobs that can be an issue. Here, 3 threads are spawned, and maybe a 4th one will be added for the [profiling signal](https://opentelemetry.io/blog/2024/profiling). I won't have any figures to share but the OTEL library could become a better citizen in the .NET ecosystem by using the thread pool.

File: src/OpenTelemetry/BaseExporter.cs
Comment (cijothomas) [inline src/OpenTelemetry/BaseExporter.cs:50]: A related discussion on this topic we had in OTel Rust community : https://github.com/open-telemetry/opentelemetry-rust/pull/2027
OTel Rust only has async API for the exporters. This is causing perf issues when the exporter does not benefit from async API (like etw exporters which has no async need as it don't talk to disk/network etc.)

We are also considering adding both options in OTel Rust, with one calling the other automatically!
Just shared as these are similar, though in diff. languages!

Comment (verdie-g) [reply src/OpenTelemetry/BaseExporter.cs:50]: The allocation overhead for calling `ExportAsync` instead of `Export` should just be the `Task` allocation. AFAIK, state machines are kept on the stack while the methods return synchronously. If a method needs to yield the execution because of a blocking operation (e.g. reading a socket), then the state machine will be moved to the heap.

In this [short snippet](https://sharplab.io/#v2:D4AQDABCCMCsDcBYAUCkBOCBBAzgTwDsBjAeQDcBTAJwGVCiAKASiVTZADZt9jzq7izFAG8UEcRAD0kiABMA9hBzyAthQAuACwCWBAOZiJIAOxQOAOgDCqgA4AbDRVmdWAXyA===) we can see that the state machine is a struct and when the async method doesn't return synchronously (i.e. in the `if (!awaiter.IsCompleted)` branch), `AwaitUnsafeOnCompleted` is called which [boxes the state machine](https://github.com/dotnet/runtime/blob/8ae3796ef109962725833610c26e3f0bb06e2988/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncTaskMethodBuilderT.cs#L92).

So I'm not very concerned about the performance impact for synchronous exporters, especially that the method should only be called every X seconds.

One thing that could be done is to use `ValueTask` to avoid the `Task` allocation. It is used for example in [Stream.ReadAsync(Memory<byte>)](https://github.com/dotnet/runtime/blob/8ae3796ef109962725833610c26e3f0bb06e2988/src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs#L308). Though `ValueTask` has more constraints than `Task` so it can be a tough decision to make. That's why I always use `Task` and it never was a bottleneck. More reading from our savior Stephen Toub: https://devblogs.microsoft.com/dotnet/understanding-the-whys-whats-and-whens-of-valuetask.

Comment (cijothomas) [reply src/OpenTelemetry/BaseExporter.cs:50]: > So I'm not very concerned about the performance impact for synchronous exporters, especially that the method should only be called every X seconds.

The export method is called for every Log, every Span when using sync exporters like ETW. (they don't do any batching)

Comment (verdie-g) [reply src/OpenTelemetry/BaseExporter.cs:50]: That could make sense to use ValueTask in that case. Could you give me a pointer to the ETW exporter. I can't seem to find it

Comment (cijothomas) [reply src/OpenTelemetry/BaseExporter.cs:50]: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/src/OpenTelemetry.Exporter.Geneva/Internal/ReentrantExportProcessor.cs - This is the exporting processor used with ETW exporters.

Comment (CodeBlanch) [inline src/OpenTelemetry/BaseExporter.cs:50]: If we are going to do this (and that is a big IF), it MUST be:

```csharp
public virtual Task<ExportResult> ExportAsync(Batch<T> batch, CancellationToken cancellationToken)
```

Because this is an opportunity to clean up a mistake 😄 That `cancellationToken` should be built from https://github.com/open-telemetry/opentelemetry-dotnet/blob/1c01770882bb5113ff308b2b4398347fab6e0404/src/OpenTelemetry/BatchExportProcessor.cs#L24C27-L24C54.

I would also like to see:

```csharp
public virtual ValueTask<ExportResult> 
```

There's only 2 `ExportResult` values. We could have 2 cached instances of `Task<ExportResult>`. But I think `ValueTask` is better for implementations which may always execute in sync. Imagine an exporter writing to a Stream. Could be async when hooked up to like a NetworkStream. But could also always run synchronously when hooked up to a MemoryStream.

**Review by CodeBlanch — CHANGES_REQUESTED** [review]
I would love to have ExportAsync but I think in general this may be the wrong direction. Our telemetry pipeline is not asynchronous. It is synchronous. And the specification says it must be so. Because of that we're always going to run into somewhere having to jump from sync to async back to sync. Async only works when everything is async. The telemetry pipeline being synchronous is a good thing. We don't want to block real work being done on a thread because it tried to write some telemetry and the SDK has to wait on a busy thread pool.

Here are some questions:

* How will `SimpleExporterProcessor` work with `ExportAsync`? Are we expecting `SimpleExporterProcessor` to continue calling `Export` or will it call `ExportAsync`? The later will probably be a non-starter. We can't tie up two threads for our telemetry pipelines.

* Depending on that answer, can we see some of the existing exporters (maybe start with Otlp) updated as a demonstration? We _require_ `Export` be implemented via `abstract`. Authors will have to code _something_ synchronous. If we offer `ExportAsync` how does that help with this? I don't want to see this...

```csharp
public override ExportResult Export(in Batch<T> batch)
{
    return this.ExportAsync(in batch).Result;
}
```

Because that will tie up two threads 👎 Or lead to an infinite loop in the event `ExportAsync` is not overridden 🤣 Also, how do we prevent either of these thing from happening?

What I'm trying to figure out is, if I'm an exporter author, and I have to do a synchronous version anyway, why would I bother with an asynchronous version?

**Comment by verdie-g** [comment]
> It is synchronous. And the specification says it must be so.

I don't think that's true. Here is a quote from [the spec](https://opentelemetry.io/docs/specs/otel/trace/sdk/#exportbatch):

> Depending on the implementation the result of the export may be returned to the Processor not in the return value of the call to Export() but in a language specific way for signaling completion of an asynchronous task.

> Because of that we're always going to run into somewhere having to jump from sync to async back to sync.

I've started implementing the `ExportAsync` method around, and found cases where we need to synchronously wait on a Task but I don't think I found a case where a synchronous operation occurs after an asynchronous one. Actually, I'm not sure what you mean. Could you clarify that point?

> The telemetry pipeline being synchronous is a good thing. We don't want to block real work being done on a thread because it tried to write some telemetry and the SDK has to wait on a busy thread pool.

When a context switch occurs to work on the OTEL thread, it will actually prevent a thread pool thread to do its job. Using the thread-pool and async exports would greatly reduce that overhead.

> How will SimpleExporterProcessor work with ExportAsync? Are we expecting SimpleExporterProcessor to continue calling Export or will it call ExportAsync?

Either should be fine. If a network exporter is used with SimpleExporterProcessor, there is obviously no guarantees about the performance, so I don't think `ExportAsync` will change anything here.

It is also worth noting that because of the synchronous export API, some exporters need to [do some sync-over-async on some platforms](https://github.com/open-telemetry/opentelemetry-dotnet/blob/1c01770882bb5113ff308b2b4398347fab6e0404/src/OpenTelemetry.Exporter.Zipkin/ZipkinExporter.cs#L80-L83).

> Depending on that answer, can we see some of the existing exporters (maybe start with Otlp) updated as a demonstration? We require Export be implemented via abstract. Authors will have to code something synchronous. If we offer ExportAsync how does that help with this? I don't want to see this...

For OTLP, the `Export` method will indeed be doing sync-over-async on the `ExportAsync`. But it does not matter because only `ExportAsync` would be called.

> What I'm trying to figure out is, if I'm an exporter author, and I have to do a synchronous version anyway, why would I bother with an asynchronous version?

I think it's very similar to how the `Stream` API works in .NET. If we focus on the read API, by default, you only need to implement ` Read(byte[] buffer, int offset, int count)`. Other methods, such as `ReadByte` or `ReadAsync` will fallback on `Read`. The stream will work fine but you'll get a performance hit until you implement the other methods.

Here is a very concrete example https://github.com/dotutils/streamutils/pull/2 where some benchmarks showed that this `Stream` was a performance bottleneck, so extra Read methods were implemented to fix that.

In the case of `ExportAsync` it will be the same. Though, it is to expect to have way less implementations of `BaseExporter` than implementations of `Stream` in the wild. Also, most of the `BaseExporter` implementations are under our control in the OTEL repositories.

File: src/OpenTelemetry/BaseExporter.cs
Comment (verdie-g) [reply src/OpenTelemetry/BaseExporter.cs]: Good idea of the CancellationToken!

Regarding the `ValueTask`, given the constraint of that type, I find it a hard decision to make. I've briefly discussed it here https://github.com/open-telemetry/opentelemetry-dotnet/pull/5838#discussion_r1763914552. I think I'll be able to give more info once I implemented ExportAsync for most exporters.

**Comment by verdie-g** [comment]
Here is what the implementations of BaseExport.ExportAsync could look like: https://github.com/verdie-g/opentelemetry-dotnet/commit/02eb4503bcd5dfcb415351f1791ddc3b913eb5a7

**Comment by verdie-g** [comment]
Kind of awkward that the PR gets automatically closed when it's waiting for a review 🤔 @CodeBlanch @cijothomas could you have a look at it again? The TL;DR is that you don't create threads in .NET. Some platforms don't even support it. And OTEL, which will be probably become the most downloaded nuget, creates 3-4 of them when it could use the thread pool.

File: src/OpenTelemetry/BatchExportProcessor.cs
Comment (cijothomas) [inline src/OpenTelemetry/BatchExportProcessor.cs:61]: I think we shouldn't change the default behavior here. If there are scenarios where a thread pool must be leveraged, then it can be a optional, opt-in behavior - either a modification for the BatchExportProcessor (UseDedicatedThread=false) or a new `BatchExportProcessorThreadPool`.

Having said that, I'll not have bandwidth to do a detailed review at the moment, but I believe the first step is to get some guidance from the maintainers about which direction should be taken:
1. Provide a opt-in feature flag to BatchExportProcessor so it can do dedicated thread or threadpool.
2. Make a new BatchExportProcessor to do the thread pool one - either in this repo OR in the contrib repo.
3. Something else.

**Review by cijothomas — COMMENTED** [review]
Left a comment about keeping current behavior as-is.

**Comment by alexaka1** [comment]
@CodeBlanch Any chance this can be reopened now?

**Comment by Kielek** [comment]
Reopen to bring @CodeBlanch attention;)

File: src/OpenTelemetry/BatchExportProcessor.cs
Comment (verdie-g) [reply src/OpenTelemetry/BatchExportProcessor.cs:61]: That's a good idea. I'll do some experimentations.

**Comment by verdie-g** [comment]
I've added a boolean to the `BatchExportProcessor` constructor so by default the behavior doesn't change. Tell me what you think and I'll add the tests.

I've first tried to create a new async implementation of `BaseExportProcessor` but that would cause a breaking change when inheriting from that one instead of `BatchExportProcessor`.

Alternatively we could also use <a href="https://learn.microsoft.com/en-us/dotnet/api/system.appcontext.trygetswitch?view=net-8.0#system-appcontext-trygetswitch(system-string-system-boolean@)">AppContext.TryGetSwitch</a> to have a global flag instead of a by-instance one.

**Comment by madskonradsen** [comment]
Still awaiting review.

**Comment by madskonradsen** [comment]
Still not stale

**Comment by alexaka1** [comment]
Afaik this PR is ready to review. Can we reopen it again @Kielek?

**Comment by alexaka1** [comment]
I really really hate to pry, but I believe this PR fell through the cracks and that's the only reason it was closed. Not because it is not needed. @CodeBlanch @Kielek

**Pull request closed without merging**
